$LAAS: patch-ab 2008/08/06 16:35:04 mallet $

   Swap memory for image filtering can be allocated in distinct banks.
    
    A swap bank number is a small integer (currently smaller than 4). Swap
    space can be allocated in distinct banks if several distinct memory areas
    are needed at the same time.

diff --git src/swap.c src/swap.c
index 9f7a664..9e0f19f 100644
--- src/swap.c
+++ src/swap.c
@@ -30,13 +30,15 @@
 
 #include <stdlib.h>
 #include <err.h>
+#include <assert.h>
 
 #include "viam_private.h"
 
 /* --- local data ---------------------------------------------------- */
 
-static void *	swap;		/**< temporary buffer for image filtering */
-static size_t	swapsize;	/**< size of swap buffer */
+static void *	swap[4];	/**< temporary buffer for image filtering */
+static size_t	swapsize[4];	/**< size of swap buffer */
+static size_t	totalsize;
 
 
 /* --- viam_swap ----------------------------------------------------- */
@@ -47,9 +49,10 @@ static size_t	swapsize;	/**< size of swap buffer */
  */
 
 void *
-viam_swap()
+viam_swap(unsigned int bank)
 {
-  return swap;
+  assert(bank < sizeof(swap)/sizeof(swap[0]));
+  return swap[bank];
 }
 
 
@@ -57,29 +60,35 @@ viam_swap()
 
 /** \brief Allocate swap buffer
  *
+ * \param[in] bank	Swap area number.
  * \param[in] size	Required amount of swap space.
  *
  * \return A pointer on the swap buffer.
  */
 
 void
-viam_setswap(size_t size)
+viam_setswap(unsigned int bank, size_t size)
 {
+  assert(bank < sizeof(swap)/sizeof(swap[0]));
+
   if (size == 0) {
-    if (swap) free(swap);
-    swap = NULL;
-    swapsize = 0;
+    if (swap[bank]) free(swap[bank]);
+    totalsize -= swapsize[bank];
+    swap[bank] = NULL;
+    swapsize[bank] = 0;
     return;
   }
-  if (size <= swapsize) return;
+  if (size <= swapsize[bank]) return;
 
-  if (swap) free(swap);
-  swap = malloc(size);
-  swapsize = size;
-  if (!swap) {
+  if (swap[bank]) free(swap[bank]);
+  swap[bank] = malloc(size);
+  totalsize -= swapsize[bank];
+  swapsize[bank] = size;
+  if (!swap[bank]) {
     warnx("cannot allocate %d bytes for camera swap", size);
-    swapsize = 0;
+    swapsize[bank] = 0;
   }
+  totalsize += swapsize[bank];
 
-  warnx("%d bytes of swap space allocated for image filtering", swapsize);
+  warnx("%d bytes of swap space allocated for image filtering", totalsize);
 }
diff --git src/viam_private.h src/viam_private.h
index ca22b52..e2eeda3 100644
--- src/viam_private.h
+++ src/viam_private.h
@@ -1,4 +1,4 @@
-/* $LAAS: viam_private.h 2008/07/05 19:32:38 tho $
+/* $LAAS: viam_private.h 2008/08/06 14:50:19 mallet $
  *
  * Copyright (c) 2007-2008 LAAS/CNRS
  * All rights reserved.
@@ -92,8 +92,8 @@ viam_cameracalib_t	viam_image_caldata(const viam_image_t image);
 void			viam_image_setcaldata(viam_image_t image,
 				viam_cameracalib_t c);
 
-void *			viam_swap(void);
-void			viam_setswap(size_t size);
+void *			viam_swap(unsigned int bank);
+void			viam_setswap(unsigned int bank, size_t size);
 
 
 /* --- Image filters ------------------------------------------------- */
